NullPointerException when annotating com.google.cloud:spring-cloud-gcp-core:jar:2.0.0-SNAPSHOT

When I tried to apply LinkageChecker to spring-cloud-gcp, it threw a NullPointerException for dependency graph of `com.google.cloud:spring-cloud-gcp-core:jar:2.0.0-SNAPSHOT`.

```
Caused by: java.lang.NullPointerException
    at com.google.cloud.tools.opensource.dependencies.DependencyGraph.levelOrder (DependencyGraph.java:259)
    at com.google.cloud.tools.opensource.dependencies.DependencyGraph.from (DependencyGraph.java:244)
    at com.google.cloud.tools.opensource.dependencies.DependencyGraphBuilder.buildDependencyGraph (DependencyGraphBuilder.java:204)
    at com.google.cloud.tools.opensource.dependencies.DependencyGraphBuilder.buildMavenDependencyGraph (DependencyGraphBuilder.java:193)
    at com.google.cloud.tools.opensource.classpath.ClassPathBuilder.resolveWithMaven (ClassPathBuilder.java:80)
    at com.google.cloud.tools.opensource.classpath.LinkageProblemCauseAnnotator.annotate (LinkageProblemCauseAnnotator.java:51)
    at com.google.cloud.tools.dependencies.enforcer.LinkageCheckerRule.execute (LinkageCheckerRule.java:216)
    at org.apache.maven.plugins.enforcer.EnforceMojo.execute (EnforceMojo.java:205)
```

<img width="1440" alt="Screen Shot 2020-08-19 at 10 38 35" src="https://user-images.githubusercontent.com/28604/90649383-a8554e00-e208-11ea-9a62-93889d5f6338.png">

<img width="1413" alt="Screen Shot 2020-08-19 at 10 41 48" src="https://user-images.githubusercontent.com/28604/90649427-b6a36a00-e208-11ea-8e84-9dce8567164e.png">


The artifact is locally available:

```
suztomo-macbookpro44% head  ~/.m2/repository/com/google/cloud/spring-cloud-gcp-core/2.0.0-SNAPSHOT/spring-cloud-gcp-core-2.0.0-SNAPSHOT.pom
<?xml version="1.0" encoding="UTF-8"?>
<project xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
		 xmlns="http://maven.apache.org/POM/4.0.0"
		 xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 https://maven.apache.org/xsd/maven-4.0.0.xsd">
	<modelVersion>4.0.0</modelVersion>
	<parent>
		<artifactId>spring-cloud-gcp</artifactId>
		<groupId>com.google.cloud</groupId>
		<version>2.0.0-SNAPSHOT</version>
	</parent>
```

# Diagnosis

Why `com.google.cloud:spring-cloud-gcp-core:jar:2.0.0-SNAPSHOT` is not resolvable? It says `org.apache.maven.model.resolution.UnresolvableModelException: Could not find artifact io.projectreactor:reactor-bom:pom:2020.0.0-M2 in  (https://repo1.maven.org/maven2/)`.

<img width="1429" alt="Screen Shot 2020-08-19 at 10 54 20" src="https://user-images.githubusercontent.com/28604/90650917-6fb67400-e20a-11ea-80dd-17854d919702.png">


The version does not exist for the POM artifact in the Maven Central. https://search.maven.org/artifact/io.projectreactor/reactor-bom. Where is it getting then? Repository element somewhere?

It's in spring-milestone repository

```
[INFO] Downloading from spring-milestones: https://repo.spring.io/milestone/io/projectreactor/reactor-bom/2020.0.0-M2/reactor-bom-2020.0.0-M2.pom
```

The cause is `ClassPathBuilder.resolveWithMaven` does not take `repository` section of the Maven project into account.

--- a/boms/integration-tests/src/test/java/com/google/cloud/MaximumLinkageErrorsTest.java
+++ b/boms/integration-tests/src/test/java/com/google/cloud/MaximumLinkageErrorsTest.java
@@ -60,7 +60,7 @@ public class MaximumLinkageErrorsTest {
     StringBuilder message = new StringBuilder("Baseline BOM: " + baselineCoordinates + "\n");
     if (!newProblems.isEmpty()) {
       message.append("Newly introduced problems:\n");
-      message.append(LinkageProblem.formatLinkageProblems(newProblems));
+      message.append(LinkageProblem.formatLinkageProblems(newProblems, null));
       Assert.fail(message.toString());
     }
   }

--- a/dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageCheckerMain.java
+++ b/dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageCheckerMain.java
@@ -91,7 +91,7 @@ class LinkageCheckerMain {
         findLinkageProblems(linkageCheckerArguments, linkageChecker);
 
     if (!linkageProblems.isEmpty()) {
-      System.out.println(LinkageProblem.formatLinkageProblems(linkageProblems));
+      System.out.println(LinkageProblem.formatLinkageProblems(linkageProblems, null));
     }
 
     return linkageProblems;
@@ -122,20 +122,7 @@ class LinkageCheckerMain {
     LinkageProblemCauseAnnotator.annotate(classPathBuilder, classPathResult, linkageProblems);
 
     if (!linkageProblems.isEmpty()) {
-      System.out.println(LinkageProblem.formatLinkageProblems(linkageProblems));
-    }
-
-    if (!linkageProblems.isEmpty()) {
-      ImmutableSet.Builder<ClassPathEntry> problematicJars = ImmutableSet.builder();
-      for (LinkageProblem linkageProblem : linkageProblems) {
-        ClassFile targetClass = linkageProblem.getTargetClass();
-        if (targetClass != null) {
-          problematicJars.add(targetClass.getClassPathEntry());
-        }
-        ClassFile sourceClassFile = linkageProblem.getSourceClass();
-        problematicJars.add(sourceClassFile.getClassPathEntry());
-      }
-      System.out.println(classPathResult.formatDependencyPaths(problematicJars.build()));
+      System.out.println(LinkageProblem.formatLinkageProblems(linkageProblems, classPathResult));
     }
 
     if (!artifactProblems.isEmpty()) {

--- a/dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageCheckerMain.java
+++ b/dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageCheckerMain.java
@@ -57,19 +57,20 @@ class LinkageCheckerMain {
         linkageCheckerArguments.printHelp();
       }
 
-      if (linkageCheckerArguments.hasInput()) { 
-        // This is non-empty if a BOM or artifacts are specified in the argument
-        ImmutableList<Artifact> artifacts = linkageCheckerArguments.getArtifacts();
+      if (linkageCheckerArguments.hasInput()) {
+        // artifactsInArguments is not empty if a BOM or artifacts are specified in the argument.
+        // If JAR files are specified, it's empty.
+        ImmutableList<Artifact> artifactsInArguments = linkageCheckerArguments.getArtifacts();
 
-        // When JAR files are specified in the argument, artifacts are empty.
         ImmutableList<ClassPathEntry> inputClassPath;
         ImmutableSet<ClassPathEntry> entryPoints;
         List<ArtifactProblem> artifactProblems = new ArrayList<>();
-        // classPathResult is kept null if JAR files are specified in the argument
+        // classPathResult and classPathBuilder are kept null if JAR files are specified in the
+        // argument
         ClassPathResult classPathResult = null;
         ClassPathBuilder classPathBuilder = null;
 
-        if (artifacts.isEmpty()) {
+        if (artifactsInArguments.isEmpty()) {
           // When JAR files are passed as arguments, classPathResult is null, because there is no need
           // to resolve Maven dependencies.
           inputClassPath = linkageCheckerArguments.getJarFiles();
@@ -79,10 +80,10 @@ class LinkageCheckerMain {
           DependencyGraphBuilder dependencyGraphBuilder =
               new DependencyGraphBuilder(linkageCheckerArguments.getMavenRepositoryUrls());
           classPathBuilder = new ClassPathBuilder(dependencyGraphBuilder);
-          classPathResult = classPathBuilder.resolve(artifacts, false);
+          classPathResult = classPathBuilder.resolve(artifactsInArguments, false);
           inputClassPath = classPathResult.getClassPath();
           artifactProblems.addAll(classPathResult.getArtifactProblems());
-          entryPoints = ImmutableSet.copyOf(inputClassPath.subList(0, artifacts.size()));
+          entryPoints = ImmutableSet.copyOf(inputClassPath.subList(0, artifactsInArguments.size()));
         }
 
         LinkageChecker linkageChecker =
@@ -100,7 +101,7 @@ class LinkageCheckerMain {
                   .collect(toImmutableSet());
         }
 
-        if (classPathResult != null) {
+        if (!artifactsInArguments.isEmpty()) {
           LinkageProblemCauseAnnotator.annotate(classPathBuilder, classPathResult, linkageProblems);
         }
 

--- a/dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageCheckerMain.java
+++ b/dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageCheckerMain.java
@@ -24,8 +24,6 @@ import com.google.common.collect.ImmutableList;
 import com.google.common.collect.ImmutableSet;
 import java.io.IOException;
 import java.nio.file.Path;
-import java.util.ArrayList;
-import java.util.List;
 import javax.xml.stream.XMLStreamException;
 import javax.xml.transform.TransformerException;
 import org.apache.commons.cli.ParseException;
@@ -62,82 +60,14 @@ class LinkageCheckerMain {
         // If JAR files are specified, it's empty.
         ImmutableList<Artifact> artifactsInArguments = linkageCheckerArguments.getArtifacts();
 
-        ImmutableList<ClassPathEntry> inputClassPath;
-        ImmutableSet<ClassPathEntry> entryPoints;
-        List<ArtifactProblem> artifactProblems = new ArrayList<>();
-        // classPathResult and classPathBuilder are kept null if JAR files are specified in the
-        // argument
-        ClassPathResult classPathResult = null;
-        ClassPathBuilder classPathBuilder = null;
-
-        if (artifactsInArguments.isEmpty()) {
-          // When JAR files are passed as arguments, classPathResult is null, because there is no need
-          // to resolve Maven dependencies.
-          inputClassPath = linkageCheckerArguments.getJarFiles();
-          entryPoints = ImmutableSet.copyOf(inputClassPath);
-        } else {
-          // When a BOM or Maven artifacts are passed as arguments, resolve the dependencies.
-          DependencyGraphBuilder dependencyGraphBuilder =
-              new DependencyGraphBuilder(linkageCheckerArguments.getMavenRepositoryUrls());
-          classPathBuilder = new ClassPathBuilder(dependencyGraphBuilder);
-          classPathResult = classPathBuilder.resolve(artifactsInArguments, false);
-          inputClassPath = classPathResult.getClassPath();
-          artifactProblems.addAll(classPathResult.getArtifactProblems());
-          entryPoints = ImmutableSet.copyOf(inputClassPath.subList(0, artifactsInArguments.size()));
-        }
-
-        LinkageChecker linkageChecker =
-            LinkageChecker.create(
-                inputClassPath, entryPoints, linkageCheckerArguments.getInputExclusionFile());
-        ImmutableSet<LinkageProblem> linkageProblems = linkageChecker.findLinkageProblems();
-
-        if (linkageCheckerArguments.getReportOnlyReachable()) {
-          ClassReferenceGraph graph = linkageChecker.getClassReferenceGraph();
-          linkageProblems =
-              linkageProblems.stream()
-                  .filter(
-                      (LinkageProblem problem) ->
-                          graph.isReachable(problem.getSourceClass().getBinaryName()))
-                  .collect(toImmutableSet());
-        }
-
-        if (!artifactsInArguments.isEmpty()) {
-          LinkageProblemCauseAnnotator.annotate(classPathBuilder, classPathResult, linkageProblems);
-        }
-
-        Path writeAsExclusionFile = linkageCheckerArguments.getOutputExclusionFile();
-        if (writeAsExclusionFile != null) {
-          ExclusionFiles.write(writeAsExclusionFile, linkageProblems);
-          System.out.println("Wrote the linkage errors as exclusion file: " + writeAsExclusionFile);
-          return;
-        }
-
+        ImmutableSet<LinkageProblem> linkageProblems =
+            artifactsInArguments.isEmpty()
+                ? runWithJarFiles(linkageCheckerArguments)
+                : runWithArtifacts(linkageCheckerArguments);
         if (!linkageProblems.isEmpty()) {
-          System.out.println(LinkageProblem.formatLinkageProblems(linkageProblems));
-        }
-
-        if (classPathResult != null && !linkageProblems.isEmpty()) {
-          ImmutableSet.Builder<ClassPathEntry> problematicJars = ImmutableSet.builder();
-          for (LinkageProblem linkageProblem : linkageProblems) {
-            ClassFile targetClass = linkageProblem.getTargetClass();
-            if (targetClass != null) {
-              problematicJars.add(targetClass.getClassPathEntry());
-            }
-            ClassFile sourceClassFile = linkageProblem.getSourceClass();
-            problematicJars.add(sourceClassFile.getClassPathEntry());
-          }
-          System.out.println(classPathResult.formatDependencyPaths(problematicJars.build()));
-        }
-    
-        if (!artifactProblems.isEmpty()) {
-          System.out.println("\n");
-          System.out.println(ArtifactProblem.formatProblems(artifactProblems));
           System.out.println(
               "For the details of the linkage errors, see "
                   + "https://github.com/GoogleCloudPlatform/cloud-opensource-java/wiki/Linkage-Checker-Messages");
-        }
-
-        if (!linkageProblems.isEmpty()) {
           // Throwing an exception is more test-friendly compared with System.exit(1). The latter
           // abruptly stops test execution.
           throw new LinkageCheckResultException(linkageProblems.size());
@@ -146,5 +76,103 @@ class LinkageCheckerMain {
     } catch (ParseException ex) {
       System.err.println(ex.getMessage());
     }
+  }
+
+  private static ImmutableSet<LinkageProblem> runWithJarFiles(
+      LinkageCheckerArguments linkageCheckerArguments)
+      throws IOException, TransformerException, XMLStreamException {
+    ImmutableList<ClassPathEntry> inputClassPath = linkageCheckerArguments.getJarFiles();
+    ImmutableSet<ClassPathEntry> entryPoints = ImmutableSet.copyOf(inputClassPath);
+    LinkageChecker linkageChecker =
+        LinkageChecker.create(
+            inputClassPath, entryPoints, linkageCheckerArguments.getInputExclusionFile());
+    ImmutableSet<LinkageProblem> linkageProblems = linkageChecker.findLinkageProblems();
+
+    if (linkageCheckerArguments.getReportOnlyReachable()) {
+      ClassReferenceGraph graph = linkageChecker.getClassReferenceGraph();
+      linkageProblems =
+          linkageProblems.stream()
+              .filter(
+                  (LinkageProblem problem) ->
+                      graph.isReachable(problem.getSourceClass().getBinaryName()))
+              .collect(toImmutableSet());
+    }
+
+    Path writeAsExclusionFile = linkageCheckerArguments.getOutputExclusionFile();
+    if (writeAsExclusionFile != null) {
+      ExclusionFiles.write(writeAsExclusionFile, linkageProblems);
+      System.out.println("Wrote the linkage errors as exclusion file: " + writeAsExclusionFile);
+      return ImmutableSet.of();
+    }
+
+    if (!linkageProblems.isEmpty()) {
+      System.out.println(LinkageProblem.formatLinkageProblems(linkageProblems));
+    }
+
+    return linkageProblems;
+  }
+
+  private static ImmutableSet<LinkageProblem> runWithArtifacts(
+      LinkageCheckerArguments linkageCheckerArguments)
+      throws IOException, RepositoryException, TransformerException, XMLStreamException {
+    ImmutableList<Artifact> artifactsInArguments = linkageCheckerArguments.getArtifacts();
+
+    // When a BOM or Maven artifacts are passed as arguments, resolve the dependencies.
+    DependencyGraphBuilder dependencyGraphBuilder =
+        new DependencyGraphBuilder(linkageCheckerArguments.getMavenRepositoryUrls());
+    ClassPathBuilder classPathBuilder = new ClassPathBuilder(dependencyGraphBuilder);
+    ClassPathResult classPathResult = classPathBuilder.resolve(artifactsInArguments, false);
+    ImmutableList<ClassPathEntry> inputClassPath = classPathResult.getClassPath();
+    ImmutableList<ArtifactProblem> artifactProblems =
+        ImmutableList.copyOf(classPathResult.getArtifactProblems());
+    ImmutableSet<ClassPathEntry> entryPoints =
+        ImmutableSet.copyOf(inputClassPath.subList(0, artifactsInArguments.size()));
+
+    LinkageChecker linkageChecker =
+        LinkageChecker.create(
+            inputClassPath, entryPoints, linkageCheckerArguments.getInputExclusionFile());
+    ImmutableSet<LinkageProblem> linkageProblems = linkageChecker.findLinkageProblems();
+
+    if (linkageCheckerArguments.getReportOnlyReachable()) {
+      ClassReferenceGraph graph = linkageChecker.getClassReferenceGraph();
+      linkageProblems =
+          linkageProblems.stream()
+              .filter(
+                  (LinkageProblem problem) ->
+                      graph.isReachable(problem.getSourceClass().getBinaryName()))
+              .collect(toImmutableSet());
+    }
+
+    LinkageProblemCauseAnnotator.annotate(classPathBuilder, classPathResult, linkageProblems);
+
+    Path writeAsExclusionFile = linkageCheckerArguments.getOutputExclusionFile();
+    if (writeAsExclusionFile != null) {
+      ExclusionFiles.write(writeAsExclusionFile, linkageProblems);
+      System.out.println("Wrote the linkage errors as exclusion file: " + writeAsExclusionFile);
+      return ImmutableSet.of();
+    }
+
+    if (!linkageProblems.isEmpty()) {
+      System.out.println(LinkageProblem.formatLinkageProblems(linkageProblems));
+    }
+
+    if (!linkageProblems.isEmpty()) {
+      ImmutableSet.Builder<ClassPathEntry> problematicJars = ImmutableSet.builder();
+      for (LinkageProblem linkageProblem : linkageProblems) {
+        ClassFile targetClass = linkageProblem.getTargetClass();
+        if (targetClass != null) {
+          problematicJars.add(targetClass.getClassPathEntry());
+        }
+        ClassFile sourceClassFile = linkageProblem.getSourceClass();
+        problematicJars.add(sourceClassFile.getClassPathEntry());
+      }
+      System.out.println(classPathResult.formatDependencyPaths(problematicJars.build()));
+    }
+
+    if (!artifactProblems.isEmpty()) {
+      System.out.println("\n");
+      System.out.println(ArtifactProblem.formatProblems(artifactProblems));
+    }
+    return linkageProblems;
   }
 }

--- a/dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageCheckerMain.java
+++ b/dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageCheckerMain.java
@@ -62,8 +62,8 @@ class LinkageCheckerMain {
 
         ImmutableSet<LinkageProblem> linkageProblems =
             artifactsInArguments.isEmpty()
-                ? runWithJarFiles(linkageCheckerArguments)
-                : runWithArtifacts(linkageCheckerArguments);
+                ? checkJarFiles(linkageCheckerArguments)
+                : checkArtifacts(linkageCheckerArguments);
         if (!linkageProblems.isEmpty()) {
           System.out.println(
               "For the details of the linkage errors, see "
@@ -78,7 +78,7 @@ class LinkageCheckerMain {
     }
   }
 
-  private static ImmutableSet<LinkageProblem> runWithJarFiles(
+  private static ImmutableSet<LinkageProblem> checkJarFiles(
       LinkageCheckerArguments linkageCheckerArguments)
       throws IOException, TransformerException, XMLStreamException {
     ImmutableList<ClassPathEntry> inputClassPath = linkageCheckerArguments.getJarFiles();
@@ -88,20 +88,9 @@ class LinkageCheckerMain {
             inputClassPath, entryPoints, linkageCheckerArguments.getInputExclusionFile());
     ImmutableSet<LinkageProblem> linkageProblems = linkageChecker.findLinkageProblems();
 
-    if (linkageCheckerArguments.getReportOnlyReachable()) {
-      ClassReferenceGraph graph = linkageChecker.getClassReferenceGraph();
-      linkageProblems =
-          linkageProblems.stream()
-              .filter(
-                  (LinkageProblem problem) ->
-                      graph.isReachable(problem.getSourceClass().getBinaryName()))
-              .collect(toImmutableSet());
-    }
+    linkageProblems = filterReachable(linkageCheckerArguments, linkageChecker, linkageProblems);
 
-    Path writeAsExclusionFile = linkageCheckerArguments.getOutputExclusionFile();
-    if (writeAsExclusionFile != null) {
-      ExclusionFiles.write(writeAsExclusionFile, linkageProblems);
-      System.out.println("Wrote the linkage errors as exclusion file: " + writeAsExclusionFile);
+    if (writeExclusionFile(linkageCheckerArguments, linkageProblems)) {
       return ImmutableSet.of();
     }
 
@@ -112,7 +101,39 @@ class LinkageCheckerMain {
     return linkageProblems;
   }
 
-  private static ImmutableSet<LinkageProblem> runWithArtifacts(
+  private static ImmutableSet<LinkageProblem> filterReachable(
+      LinkageCheckerArguments linkageCheckerArguments,
+      LinkageChecker linkageChecker,
+      ImmutableSet<LinkageProblem> linkageProblems) {
+    if (linkageCheckerArguments.getReportOnlyReachable()) {
+      ClassReferenceGraph graph = linkageChecker.getClassReferenceGraph();
+      linkageProblems =
+          linkageProblems.stream()
+              .filter(
+                  (LinkageProblem problem) ->
+                      graph.isReachable(problem.getSourceClass().getBinaryName()))
+              .collect(toImmutableSet());
+    }
+    return linkageProblems;
+  }
+
+  /**
+   * Writes {@code linkageProblems} to a file and returns true if {@code linkageCheckerArguments}
+   * specifies the option; otherwise returns false.
+   */
+  private static boolean writeExclusionFile(
+      LinkageCheckerArguments linkageCheckerArguments, ImmutableSet<LinkageProblem> linkageProblems)
+      throws TransformerException, XMLStreamException, IOException {
+    Path writeAsExclusionFile = linkageCheckerArguments.getOutputExclusionFile();
+    if (writeAsExclusionFile != null) {
+      ExclusionFiles.write(writeAsExclusionFile, linkageProblems);
+      System.out.println("Wrote the linkage errors as exclusion file: " + writeAsExclusionFile);
+      return true;
+    }
+    return false;
+  }
+
+  private static ImmutableSet<LinkageProblem> checkArtifacts(
       LinkageCheckerArguments linkageCheckerArguments)
       throws IOException, RepositoryException, TransformerException, XMLStreamException {
     ImmutableList<Artifact> artifactsInArguments = linkageCheckerArguments.getArtifacts();
@@ -133,22 +154,11 @@ class LinkageCheckerMain {
             inputClassPath, entryPoints, linkageCheckerArguments.getInputExclusionFile());
     ImmutableSet<LinkageProblem> linkageProblems = linkageChecker.findLinkageProblems();
 
-    if (linkageCheckerArguments.getReportOnlyReachable()) {
-      ClassReferenceGraph graph = linkageChecker.getClassReferenceGraph();
-      linkageProblems =
-          linkageProblems.stream()
-              .filter(
-                  (LinkageProblem problem) ->
-                      graph.isReachable(problem.getSourceClass().getBinaryName()))
-              .collect(toImmutableSet());
-    }
+    linkageProblems = filterReachable(linkageCheckerArguments, linkageChecker, linkageProblems);
 
     LinkageProblemCauseAnnotator.annotate(classPathBuilder, classPathResult, linkageProblems);
 
-    Path writeAsExclusionFile = linkageCheckerArguments.getOutputExclusionFile();
-    if (writeAsExclusionFile != null) {
-      ExclusionFiles.write(writeAsExclusionFile, linkageProblems);
-      System.out.println("Wrote the linkage errors as exclusion file: " + writeAsExclusionFile);
+    if (writeExclusionFile(linkageCheckerArguments, linkageProblems)) {
       return ImmutableSet.of();
     }
 

--- a/dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageCheckerMain.java
+++ b/dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageCheckerMain.java
@@ -119,7 +119,7 @@ class LinkageCheckerMain {
 
   /**
    * Writes {@code linkageProblems} to a file and returns true if {@code linkageCheckerArguments}
-   * specifies the option; otherwise returns false.
+   * has the option; otherwise returns false.
    */
   private static boolean writeExclusionFile(
       LinkageCheckerArguments linkageCheckerArguments, ImmutableSet<LinkageProblem> linkageProblems)

--- a/dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageCheckerMain.java
+++ b/dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageCheckerMain.java
@@ -67,7 +67,8 @@ class LinkageCheckerMain {
         List<ArtifactProblem> artifactProblems = new ArrayList<>();
         // classPathResult is kept null if JAR files are specified in the argument
         ClassPathResult classPathResult = null;
-    
+
+        ClassPathBuilder classPathBuilder = null;
         if (artifacts.isEmpty()) {
           // When JAR files are passed as arguments, classPathResult is null, because there is no need
           // to resolve Maven dependencies.
@@ -77,7 +78,7 @@ class LinkageCheckerMain {
           // When a BOM or Maven artifacts are passed as arguments, resolve the dependencies.
           DependencyGraphBuilder dependencyGraphBuilder =
               new DependencyGraphBuilder(linkageCheckerArguments.getMavenRepositoryUrls());
-          ClassPathBuilder classPathBuilder = new ClassPathBuilder(dependencyGraphBuilder);
+          classPathBuilder = new ClassPathBuilder(dependencyGraphBuilder);
           classPathResult = classPathBuilder.resolve(artifacts, false);
           inputClassPath = classPathResult.getClassPath();
           artifactProblems.addAll(classPathResult.getArtifactProblems());
@@ -100,7 +101,7 @@ class LinkageCheckerMain {
         }
 
         if (classPathResult != null) {
-          LinkageProblemCauseAnnotator.annotate(classPathResult, linkageProblems);
+          LinkageProblemCauseAnnotator.annotate(classPathBuilder, classPathResult, linkageProblems);
         }
 
         Path writeAsExclusionFile = linkageCheckerArguments.getOutputExclusionFile();

--- a/dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageCheckerMain.java
+++ b/dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageCheckerMain.java
@@ -86,51 +86,15 @@ class LinkageCheckerMain {
     LinkageChecker linkageChecker =
         LinkageChecker.create(
             inputClassPath, entryPoints, linkageCheckerArguments.getInputExclusionFile());
-    ImmutableSet<LinkageProblem> linkageProblems = linkageChecker.findLinkageProblems();
 
-    linkageProblems = filterReachable(linkageCheckerArguments, linkageChecker, linkageProblems);
-
-    if (writeExclusionFile(linkageCheckerArguments, linkageProblems)) {
-      return ImmutableSet.of();
-    }
+    ImmutableSet<LinkageProblem> linkageProblems =
+        findLinkageProblems(linkageCheckerArguments, linkageChecker);
 
     if (!linkageProblems.isEmpty()) {
       System.out.println(LinkageProblem.formatLinkageProblems(linkageProblems));
     }
 
     return linkageProblems;
-  }
-
-  private static ImmutableSet<LinkageProblem> filterReachable(
-      LinkageCheckerArguments linkageCheckerArguments,
-      LinkageChecker linkageChecker,
-      ImmutableSet<LinkageProblem> linkageProblems) {
-    if (linkageCheckerArguments.getReportOnlyReachable()) {
-      ClassReferenceGraph graph = linkageChecker.getClassReferenceGraph();
-      linkageProblems =
-          linkageProblems.stream()
-              .filter(
-                  (LinkageProblem problem) ->
-                      graph.isReachable(problem.getSourceClass().getBinaryName()))
-              .collect(toImmutableSet());
-    }
-    return linkageProblems;
-  }
-
-  /**
-   * Writes {@code linkageProblems} to a file and returns true if {@code linkageCheckerArguments}
-   * has the option; otherwise returns false.
-   */
-  private static boolean writeExclusionFile(
-      LinkageCheckerArguments linkageCheckerArguments, ImmutableSet<LinkageProblem> linkageProblems)
-      throws TransformerException, XMLStreamException, IOException {
-    Path writeAsExclusionFile = linkageCheckerArguments.getOutputExclusionFile();
-    if (writeAsExclusionFile != null) {
-      ExclusionFiles.write(writeAsExclusionFile, linkageProblems);
-      System.out.println("Wrote the linkage errors as exclusion file: " + writeAsExclusionFile);
-      return true;
-    }
-    return false;
   }
 
   private static ImmutableSet<LinkageProblem> checkArtifacts(
@@ -152,15 +116,10 @@ class LinkageCheckerMain {
     LinkageChecker linkageChecker =
         LinkageChecker.create(
             inputClassPath, entryPoints, linkageCheckerArguments.getInputExclusionFile());
-    ImmutableSet<LinkageProblem> linkageProblems = linkageChecker.findLinkageProblems();
-
-    linkageProblems = filterReachable(linkageCheckerArguments, linkageChecker, linkageProblems);
+    ImmutableSet<LinkageProblem> linkageProblems =
+        findLinkageProblems(linkageCheckerArguments, linkageChecker);
 
     LinkageProblemCauseAnnotator.annotate(classPathBuilder, classPathResult, linkageProblems);
-
-    if (writeExclusionFile(linkageCheckerArguments, linkageProblems)) {
-      return ImmutableSet.of();
-    }
 
     if (!linkageProblems.isEmpty()) {
       System.out.println(LinkageProblem.formatLinkageProblems(linkageProblems));
@@ -183,6 +142,32 @@ class LinkageCheckerMain {
       System.out.println("\n");
       System.out.println(ArtifactProblem.formatProblems(artifactProblems));
     }
+    return linkageProblems;
+  }
+
+  private static ImmutableSet<LinkageProblem> findLinkageProblems(
+      LinkageCheckerArguments linkageCheckerArguments, LinkageChecker linkageChecker)
+      throws IOException, TransformerException, XMLStreamException {
+
+    ImmutableSet<LinkageProblem> linkageProblems = linkageChecker.findLinkageProblems();
+
+    if (linkageCheckerArguments.getReportOnlyReachable()) {
+      ClassReferenceGraph graph = linkageChecker.getClassReferenceGraph();
+      linkageProblems =
+          linkageProblems.stream()
+              .filter(
+                  (LinkageProblem problem) ->
+                      graph.isReachable(problem.getSourceClass().getBinaryName()))
+              .collect(toImmutableSet());
+    }
+
+    Path outputExclusionFile = linkageCheckerArguments.getOutputExclusionFile();
+    if (outputExclusionFile != null) {
+      ExclusionFiles.write(outputExclusionFile, linkageProblems);
+      System.out.println("Wrote the linkage errors as exclusion file: " + outputExclusionFile);
+      return ImmutableSet.of();
+    }
+
     return linkageProblems;
   }
 }

--- a/dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageCheckerMain.java
+++ b/dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageCheckerMain.java
@@ -67,8 +67,8 @@ class LinkageCheckerMain {
         List<ArtifactProblem> artifactProblems = new ArrayList<>();
         // classPathResult is kept null if JAR files are specified in the argument
         ClassPathResult classPathResult = null;
-
         ClassPathBuilder classPathBuilder = null;
+
         if (artifacts.isEmpty()) {
           // When JAR files are passed as arguments, classPathResult is null, because there is no need
           // to resolve Maven dependencies.

--- a/dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageProblem.java
+++ b/dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageProblem.java
@@ -152,8 +152,13 @@ public abstract class LinkageProblem {
     return ImmutableMap.copyOf(valueTransformed);
   }
 
-  /** Returns the formatted {@code linkageProblems} by grouping them by the {@code symbol}s. */
-  public static String formatLinkageProblems(Set<LinkageProblem> linkageProblems) {
+  /**
+   * Returns the formatted {@code linkageProblems} by grouping them by the {@code symbol}s. If
+   * {@code classPathResult} is not null, it supplies dependency paths from the root to the
+   * artifacts in the problems.
+   */
+  public static String formatLinkageProblems(
+      Set<LinkageProblem> linkageProblems, @Nullable ClassPathResult classPathResult) {
     StringBuilder output = new StringBuilder();
 
     // Don't group AbstractMethodProblems by symbols because they do not fit in the
@@ -215,7 +220,28 @@ public abstract class LinkageProblem {
       output.append("    " + causeWithIndent + "\n");
     }
 
+    if (classPathResult != null) {
+      String dependencyPaths = dependencyPathsOfProblematicJars(classPathResult, linkageProblems);
+      output.append(dependencyPaths);
+    }
+
     return output.toString();
+  }
+
+  private static String dependencyPathsOfProblematicJars(
+      ClassPathResult classPathResult, Set<LinkageProblem> linkageProblems) {
+    ImmutableSet.Builder<ClassPathEntry> problematicJars = ImmutableSet.builder();
+    for (LinkageProblem problem : linkageProblems) {
+      if (problem.getTargetClass() != null) {
+        problematicJars.add(problem.getTargetClass().getClassPathEntry());
+      }
+
+      ClassFile sourceClass = problem.getSourceClass();
+      problematicJars.add(sourceClass.getClassPathEntry());
+    }
+
+    return "Problematic artifacts in the dependency tree:\n"
+        + classPathResult.formatDependencyPaths(problematicJars.build());
   }
 
   String describe(DependencyConflict conflict) {

--- a/dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageProblemCauseAnnotator.java
+++ b/dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageProblemCauseAnnotator.java
@@ -30,14 +30,19 @@ public final class LinkageProblemCauseAnnotator {
   /**
    * Annotates the cause field of {@link LinkageProblem}s with the {@link LinkageProblemCause}.
    *
+   * @param classPathBuilder class path builder to resolve dependency graphs
    * @param rootResult the class path used for generating the linkage problems
    * @param linkageProblems linkage problems to annotate
    * @throws IOException when there is a problem reading JAR files
    */
-  public static void annotate(ClassPathResult rootResult, Iterable<LinkageProblem> linkageProblems)
+  public static void annotate(
+      ClassPathBuilder classPathBuilder,
+      ClassPathResult rootResult,
+      Iterable<LinkageProblem> linkageProblems)
       throws IOException {
 
     Map<Artifact, ClassPathResult> cache = new HashMap<>();
+
     for (LinkageProblem linkageProblem : linkageProblems) {
       ClassFile sourceClass = linkageProblem.getSourceClass();
       ClassPathEntry sourceEntry = sourceClass.getClassPathEntry();
@@ -47,7 +52,6 @@ public final class LinkageProblemCauseAnnotator {
       ClassPathResult subtreeResult = cache.get(sourceArtifact);
       if (subtreeResult == null) {
         // Resolves the dependency graph with the source artifact at the root.
-        ClassPathBuilder classPathBuilder = new ClassPathBuilder();
         subtreeResult = classPathBuilder.resolveWithMaven(sourceArtifact);
         cache.put(sourceArtifact, subtreeResult);
       }

--- a/dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageProblemCauseAnnotator.java
+++ b/dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageProblemCauseAnnotator.java
@@ -16,6 +16,8 @@
 
 package com.google.cloud.tools.opensource.classpath;
 
+import static com.google.common.base.Preconditions.checkNotNull;
+
 import com.google.cloud.tools.opensource.dependencies.DependencyPath;
 import java.io.IOException;
 import java.util.HashMap;
@@ -40,9 +42,11 @@ public final class LinkageProblemCauseAnnotator {
       ClassPathResult rootResult,
       Iterable<LinkageProblem> linkageProblems)
       throws IOException {
+    checkNotNull(classPathBuilder);
+    checkNotNull(rootResult);
+    checkNotNull(linkageProblems);
 
     Map<Artifact, ClassPathResult> cache = new HashMap<>();
-
     for (LinkageProblem linkageProblem : linkageProblems) {
       ClassFile sourceClass = linkageProblem.getSourceClass();
       ClassPathEntry sourceEntry = sourceClass.getClassPathEntry();

--- a/dependencies/src/main/java/com/google/cloud/tools/opensource/dependencies/Java8IncompatibleReferenceCheck.java
+++ b/dependencies/src/main/java/com/google/cloud/tools/opensource/dependencies/Java8IncompatibleReferenceCheck.java
@@ -105,7 +105,9 @@ public class Java8IncompatibleReferenceCheck {
             .map(ClassPathEntry::getArtifact)
             .forEach(artifact -> problematicDependencies.put(managedDependency, artifact));
 
-        logger.severe(LinkageProblem.formatLinkageProblems(invalidReferencesToJavaCoreLibrary));
+        // No need to supply classPath result as the artifact information is output below
+        logger.severe(
+            LinkageProblem.formatLinkageProblems(invalidReferencesToJavaCoreLibrary, null));
       }
     }
 

--- a/dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/LinkageProblemCauseAnnotatorTest.java
+++ b/dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/LinkageProblemCauseAnnotatorTest.java
@@ -19,6 +19,7 @@ package com.google.cloud.tools.opensource.classpath;
 import static org.junit.Assert.assertEquals;
 import static org.junit.Assert.assertTrue;
 
+import com.google.cloud.tools.opensource.dependencies.Artifacts;
 import com.google.cloud.tools.opensource.dependencies.DependencyGraphBuilder;
 import com.google.cloud.tools.opensource.dependencies.DependencyPath;
 import com.google.cloud.tools.opensource.dependencies.RepositoryUtility;
@@ -154,12 +155,14 @@ public class LinkageProblemCauseAnnotatorTest {
     ClassPathBuilder classPathBuilderWithSpring = new ClassPathBuilder(dependencyGraphBuilder);
     ClassPathResult classPathResult =
         classPathBuilderWithSpring.resolve(
-            ImmutableList.of(new DefaultArtifact("io.projectreactor:reactor-core:3.4.0-M2")),
+            ImmutableList.of(
+                new DefaultArtifact("io.projectreactor:reactor-core:3.4.0-M2"),
+                new DefaultArtifact("org.reactivestreams:reactive-streams:1.0.0")),
             false);
 
     ClassPathEntry reactorCore = classPathResult.getClassPath().get(0);
 
-    // A hypothetical problem when reactor.util.Loggers class is missing
+    // A hypothetical problem where org.reactivestreams.Subscriber class is missing
     LinkageProblem problem =
         new ClassNotFoundProblem(
             new ClassFile(reactorCore, "reactor.util.Metrics"),
@@ -170,6 +173,17 @@ public class LinkageProblemCauseAnnotatorTest {
         classPathBuilderWithSpring, classPathResult, ImmutableSet.of(problem));
 
     LinkageProblemCause cause = problem.getCause();
-    assertEquals(UnknownCause.class, cause.getClass());
+    assertEquals(DependencyConflict.class, cause.getClass());
+    DependencyConflict conflict = (DependencyConflict) cause;
+
+    // The class path has reactive-streams:1.0.0
+    assertEquals(
+        "org.reactivestreams:reactive-streams:1.0.0",
+        Artifacts.toCoordinates(conflict.getPathToSelectedArtifact().getLeaf()));
+
+    // io.projectreactor:reactor-core depends on reactive-streams 1.0.3
+    assertEquals(
+        "org.reactivestreams:reactive-streams:1.0.3",
+        Artifacts.toCoordinates(conflict.getPathToArtifactThruSource().getLeaf()));
   }
 }

--- a/dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/LinkageProblemCauseAnnotatorTest.java
+++ b/dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/LinkageProblemCauseAnnotatorTest.java
@@ -153,16 +153,19 @@ public class LinkageProblemCauseAnnotatorTest {
                 "https://repo.spring.io/milestone", RepositoryUtility.CENTRAL.getUrl()));
 
     ClassPathBuilder classPathBuilderWithSpring = new ClassPathBuilder(dependencyGraphBuilder);
+
+    // io.projectreactor:reactor-core:3.4.0-M2 is in the Spring Milestones repository.
     ClassPathResult classPathResult =
         classPathBuilderWithSpring.resolve(
             ImmutableList.of(
                 new DefaultArtifact("io.projectreactor:reactor-core:3.4.0-M2"),
-                new DefaultArtifact("org.reactivestreams:reactive-streams:1.0.0")),
+                new DefaultArtifact("org.reactivestreams:reactive-streams:0.4.0")),
             false);
 
     ClassPathEntry reactorCore = classPathResult.getClassPath().get(0);
 
-    // A hypothetical problem where org.reactivestreams.Subscriber class is missing
+    // A hypothetical problem where org.reactivestreams.Subscriber class is missing because the
+    // org.reactivestreams:reactive-streams, which contains the class, has a different version.
     LinkageProblem problem =
         new ClassNotFoundProblem(
             new ClassFile(reactorCore, "reactor.util.Metrics"),
@@ -176,9 +179,8 @@ public class LinkageProblemCauseAnnotatorTest {
     assertEquals(DependencyConflict.class, cause.getClass());
     DependencyConflict conflict = (DependencyConflict) cause;
 
-    // The class path has reactive-streams:1.0.0
     assertEquals(
-        "org.reactivestreams:reactive-streams:1.0.0",
+        "org.reactivestreams:reactive-streams:0.4.0",
         Artifacts.toCoordinates(conflict.getPathToSelectedArtifact().getLeaf()));
 
     // io.projectreactor:reactor-core depends on reactive-streams 1.0.3

--- a/dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/LinkageProblemCauseAnnotatorTest.java
+++ b/dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/LinkageProblemCauseAnnotatorTest.java
@@ -19,7 +19,9 @@ package com.google.cloud.tools.opensource.classpath;
 import static org.junit.Assert.assertEquals;
 import static org.junit.Assert.assertTrue;
 
+import com.google.cloud.tools.opensource.dependencies.DependencyGraphBuilder;
 import com.google.cloud.tools.opensource.dependencies.DependencyPath;
+import com.google.cloud.tools.opensource.dependencies.RepositoryUtility;
 import com.google.common.collect.ImmutableList;
 import com.google.common.collect.ImmutableSet;
 import java.io.IOException;
@@ -29,6 +31,8 @@ import org.eclipse.aether.artifact.DefaultArtifact;
 import org.junit.Test;
 
 public class LinkageProblemCauseAnnotatorTest {
+
+  private ClassPathBuilder classPathBuilder = new ClassPathBuilder();
 
   @Test
   public void testAnnotate_dom4jOptionalDependency() throws IOException {
@@ -48,7 +52,8 @@ public class LinkageProblemCauseAnnotatorTest {
             new ClassFile(dom4jEntry, "org.dom4j.DocumentHelper"),
             new ClassSymbol("org.jaxen.VariableContext"));
 
-    LinkageProblemCauseAnnotator.annotate(classPathResult, ImmutableSet.of(problem));
+    LinkageProblemCauseAnnotator.annotate(
+        classPathBuilder, classPathResult, ImmutableSet.of(problem));
 
     LinkageProblemCause cause = problem.getCause();
     assertEquals(MissingDependency.class, cause.getClass());
@@ -81,7 +86,8 @@ public class LinkageProblemCauseAnnotatorTest {
     SymbolNotFoundProblem problem = (SymbolNotFoundProblem) foundProblem.get();
     assertEquals("verify", ((MethodSymbol) problem.getSymbol()).getName());
 
-    LinkageProblemCauseAnnotator.annotate(classPathResult, ImmutableSet.of(problem));
+    LinkageProblemCauseAnnotator.annotate(
+        classPathBuilder, classPathResult, ImmutableSet.of(problem));
 
     LinkageProblemCause cause = problem.getCause();
     assertTrue(cause instanceof DependencyConflict);
@@ -125,7 +131,8 @@ public class LinkageProblemCauseAnnotatorTest {
                 autoServiceEntry, "com.google.auto.service.processor.AutoServiceProcessor"),
             new ClassSymbol("com.google.auto.service.AutoService"));
 
-    LinkageProblemCauseAnnotator.annotate(classPathResult, ImmutableSet.of(problem));
+    LinkageProblemCauseAnnotator.annotate(
+        classPathBuilder, classPathResult, ImmutableSet.of(problem));
 
     LinkageProblemCause cause = problem.getCause();
     assertEquals(ExcludedDependency.class, cause.getClass());
@@ -135,5 +142,34 @@ public class LinkageProblemCauseAnnotatorTest {
     assertEquals("auto-service-annotations", leaf.getArtifactId());
 
     assertEquals("auto-value", excludedDependency.getExcludingArtifact().getArtifactId());
+  }
+
+  @Test
+  public void testAnnotate_dependencyInSpringRepository() throws IOException {
+    DependencyGraphBuilder dependencyGraphBuilder =
+        new DependencyGraphBuilder(
+            ImmutableList.of(
+                "https://repo.spring.io/milestone", RepositoryUtility.CENTRAL.getUrl()));
+
+    ClassPathBuilder classPathBuilderWithSpring = new ClassPathBuilder(dependencyGraphBuilder);
+    ClassPathResult classPathResult =
+        classPathBuilderWithSpring.resolve(
+            ImmutableList.of(new DefaultArtifact("io.projectreactor:reactor-core:3.4.0-M2")),
+            false);
+
+    ClassPathEntry reactorCore = classPathResult.getClassPath().get(0);
+
+    // A hypothetical problem when reactor.util.Loggers class is missing
+    LinkageProblem problem =
+        new ClassNotFoundProblem(
+            new ClassFile(reactorCore, "reactor.util.Metrics"),
+            new ClassSymbol("org.reactivestreams.Subscriber"));
+
+    // This should not throw exception in resolving dependency graph
+    LinkageProblemCauseAnnotator.annotate(
+        classPathBuilderWithSpring, classPathResult, ImmutableSet.of(problem));
+
+    LinkageProblemCause cause = problem.getCause();
+    assertEquals(UnknownCause.class, cause.getClass());
   }
 }

--- a/dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/LinkageProblemTest.java
+++ b/dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/LinkageProblemTest.java
@@ -155,7 +155,7 @@ public class LinkageProblemTest {
             + "(com.google:ccc:1.2.3) java.lang.Integer's field MAX_VALUE is not found;\n"
             + "  referenced by 1 class file\n"
             + "    java.lang.Integer (com.google:bar:0.0.1)\n",
-        LinkageProblem.formatLinkageProblems(linkageProblems));
+        LinkageProblem.formatLinkageProblems(linkageProblems, null));
   }
 
   @Test

--- a/enforcer-rules/src/it/bom-project-error/verify.groovy
+++ b/enforcer-rules/src/it/bom-project-error/verify.groovy
@@ -28,7 +28,7 @@ Class org.apache.log.Logger is not found;
       unselected: io.grpc:grpc-core:1.17.1 (compile) / com.google.guava:guava:26.0-android (compile)
 ''')
 
-assert buildLog.contains('''[ERROR] Problematic artifacts in the dependency tree:
+assert buildLog.contains('''Problematic artifacts in the dependency tree:
 commons-logging:commons-logging:1.1.1 is at:
   com.google.api-client:google-api-client:1.27.0 (compile) / com.google.oauth-client:google-oauth-client:1.27.0 (compile) / com.google.http-client:google-http-client:1.27.0 (compile) / com.google.android:android:1.5_r4 (provided) / commons-logging:commons-logging:1.1.1 (compile)
   and 3 other dependency paths.

--- a/enforcer-rules/src/it/fail-build-for-linkage-errors/verify.groovy
+++ b/enforcer-rules/src/it/fail-build-for-linkage-errors/verify.groovy
@@ -20,7 +20,7 @@ assert buildLog.contains('''\
 ''')
 
 assert buildLog.contains('''\
-[ERROR] Problematic artifacts in the dependency tree:
+Problematic artifacts in the dependency tree:
 commons-logging:commons-logging:1.1.1 is at:
   com.google.cloud.tools.opensource:test-no-such-method-error-example:jar:1.0-SNAPSHOT \
 / com.google.api-client:google-api-client:1.27.0 (compile) \

--- a/enforcer-rules/src/main/java/com/google/cloud/tools/dependencies/enforcer/LinkageCheckerRule.java
+++ b/enforcer-rules/src/main/java/com/google/cloud/tools/dependencies/enforcer/LinkageCheckerRule.java
@@ -20,7 +20,6 @@ import static com.google.common.collect.ImmutableList.toImmutableList;
 import static com.google.common.collect.ImmutableSet.toImmutableSet;
 import static org.apache.maven.enforcer.rule.api.EnforcerLevel.WARN;
 
-import com.google.cloud.tools.opensource.classpath.ClassFile;
 import com.google.cloud.tools.opensource.classpath.ClassPathBuilder;
 import com.google.cloud.tools.opensource.classpath.ClassPathEntry;
 import com.google.cloud.tools.opensource.classpath.ClassPathResult;
@@ -236,16 +235,14 @@ public class LinkageCheckerRule extends AbstractNonCacheableEnforcerRule {
           String message =
               String.format(
                   "Linkage Checker rule found %d %s. Linkage error report:\n%s",
-                  errorCount, foundError, LinkageProblem.formatLinkageProblems(linkageProblems));
-          String dependencyPaths =
-              dependencyPathsOfProblematicJars(classPathResult, linkageProblems);
+                  errorCount,
+                  foundError,
+                  LinkageProblem.formatLinkageProblems(linkageProblems, classPathResult));
 
           if (getLevel() == WARN) {
             logger.warn(message);
-            logger.warn(dependencyPaths);
           } else {
             logger.error(message);
-            logger.error(dependencyPaths);
             logger.info(
                 "For the details of the linkage errors, see "
                     + "https://github.com/GoogleCloudPlatform/cloud-opensource-java/wiki/Linkage-Checker-Messages");
@@ -384,21 +381,5 @@ public class LinkageCheckerRule extends AbstractNonCacheableEnforcerRule {
       throw new EnforcerRuleException("Failed to collect dependency: " + artifactProblems);
     }
     return result;
-  }
-
-  private String dependencyPathsOfProblematicJars(
-      ClassPathResult classPathResult, Set<LinkageProblem> linkageProblems) {
-    ImmutableSet.Builder<ClassPathEntry> problematicJars = ImmutableSet.builder();
-    for (LinkageProblem problem : linkageProblems) {
-      if (problem.getTargetClass() != null) {
-        problematicJars.add(problem.getTargetClass().getClassPathEntry());
-      }
-
-      ClassFile sourceClass = problem.getSourceClass();
-      problematicJars.add(sourceClass.getClassPathEntry());
-    }
-
-    return "Problematic artifacts in the dependency tree:\n"
-        + classPathResult.formatDependencyPaths(problematicJars.build());
   }
 }

--- a/enforcer-rules/src/main/java/com/google/cloud/tools/dependencies/enforcer/LinkageCheckerRule.java
+++ b/enforcer-rules/src/main/java/com/google/cloud/tools/dependencies/enforcer/LinkageCheckerRule.java
@@ -221,7 +221,7 @@ public class LinkageCheckerRule extends AbstractNonCacheableEnforcerRule {
         }
 
         if (classPathResult != null) {
-          LinkageProblemCauseAnnotator.annotate(classPathResult, linkageProblems);
+          LinkageProblemCauseAnnotator.annotate(classPathBuilder, classPathResult, linkageProblems);
         }
 
         // Count unique LinkageProblems by their symbols

--- a/enforcer-rules/src/main/java/com/google/cloud/tools/dependencies/enforcer/LinkageCheckerRule.java
+++ b/enforcer-rules/src/main/java/com/google/cloud/tools/dependencies/enforcer/LinkageCheckerRule.java
@@ -235,7 +235,9 @@ public class LinkageCheckerRule extends AbstractNonCacheableEnforcerRule {
           String message =
               String.format(
                   "Linkage Checker rule found %d %s:\n%s",
-                  errorCount, foundError, LinkageProblem.formatLinkageProblems(linkageProblems, classPathResult));
+                  errorCount,
+                  foundError,
+                  LinkageProblem.formatLinkageProblems(linkageProblems, classPathResult));
           if (getLevel() == WARN) {
             logger.warn(message);
           } else {

--- a/enforcer-rules/src/test/java/com/google/cloud/tools/dependencies/enforcer/LinkageCheckerRuleTest.java
+++ b/enforcer-rules/src/test/java/com/google/cloud/tools/dependencies/enforcer/LinkageCheckerRuleTest.java
@@ -36,7 +36,6 @@ import java.net.URISyntaxException;
 import java.net.URLClassLoader;
 import java.nio.file.Paths;
 import java.util.Arrays;
-import java.util.List;
 import java.util.Map;
 import java.util.Objects;
 import java.util.stream.Collectors;
@@ -232,12 +231,12 @@ public class LinkageCheckerRuleTest {
     } catch (EnforcerRuleException ex) {
       // pass
       ArgumentCaptor<String> errorMessageCaptor = ArgumentCaptor.forClass(String.class);
-      verify(mockLog, times(2)).error(errorMessageCaptor.capture());
+      verify(mockLog, times(1)).error(errorMessageCaptor.capture());
 
-      List<String> errorMessages = errorMessageCaptor.getAllValues();
-      Truth.assertThat(errorMessages.get(0)).startsWith("Linkage Checker rule found 112 errors.");
-      Truth.assertThat(errorMessages.get(1))
-          .startsWith(
+      String errorMessage = errorMessageCaptor.getValue();
+      Truth.assertThat(errorMessage).startsWith("Linkage Checker rule found 112 errors.");
+      Truth.assertThat(errorMessage)
+          .contains(
               "Problematic artifacts in the dependency tree:\n"
                   + "com.google.appengine:appengine-api-1.0-sdk:1.9.64 is at:\n"
                   + "  a:b:jar:0.1 / com.google.appengine:appengine-api-1.0-sdk:1.9.64 (compile)");

--- a/enforcer-rules/src/test/java/com/google/cloud/tools/dependencies/enforcer/LinkageCheckerRuleTest.java
+++ b/enforcer-rules/src/test/java/com/google/cloud/tools/dependencies/enforcer/LinkageCheckerRuleTest.java
@@ -234,7 +234,7 @@ public class LinkageCheckerRuleTest {
       verify(mockLog, times(1)).error(errorMessageCaptor.capture());
 
       String errorMessage = errorMessageCaptor.getValue();
-      Truth.assertThat(errorMessage).startsWith("Linkage Checker rule found 112 errors.");
+      Truth.assertThat(errorMessage).startsWith("Linkage Checker rule found 112 errors:");
       Truth.assertThat(errorMessage)
           .contains(
               "Problematic artifacts in the dependency tree:\n"

